resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala

```diff
@@ -770,6 +771,40 @@ abstract class YarnShuffleServiceSuite extends SparkFunSuite with Matchers {
     s1.stop()
   }
 
+  test("SPARK-40186: shuffleMergeManager should have been shutdown before db closed") {
+    val maxId = 100
+    s1 = createYarnShuffleService()
+    val resolver = s1.shuffleMergeManager.asInstanceOf[RemoteBlockPushResolver]
+    (0 to maxId).foreach { id =>
+      val appId = ApplicationId.newInstance(0, id)
+      val appInfo = makeAppInfo("user", appId)
+      s1.initializeApplication(appInfo)
+      val mergedShuffleInfo =
+        new ExecutorShuffleInfo(
+          Array(new File(tempDir, "foo/foo").getAbsolutePath,
+            new File(tempDir, "bar/bar").getAbsolutePath), 3,
+          SORT_MANAGER_WITH_MERGE_SHUFFLE_META_WithAttemptID1)
+      resolver.registerExecutor(appId.toString, mergedShuffleInfo)
+    }
+
+    (0 until maxId).foreach { id =>
+      val appId = ApplicationId.newInstance(0, id)
+      resolver.applicationRemoved(appId.toString, true)
+    }
+
+    s1.stop()
+
+    assert(ShuffleTestAccessor.isMergedShuffleCleanerShutdown(resolver))
+    assert(ShuffleTestAccessor.mergeManagerLevelDB(resolver) == null)
+
+    val message = intercept[RejectedExecutionException] {
```

Comment:
> It is better to keep this `final` - any particular reason to change this ?

Due to this test scene.

If not reset `db=null` after db close, then directly invoke ` resolver .applicationRemoved` will still cause `delete data from a closed db instance`(I manual tested RocksDB. If keep final and  not set  `db = null` after db closed, it will still crash.).

https://github.com/apache/spark/blob/4b0c3bab1ab082565a051990bf45774f15962deb/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L373-L394

As above code, both `removeAppAttemptPathInfoFromDB` and `removeOldApplicationAttemptsFromDb` are synchronous operations. This operation will be rejected only when db is `null`.

Therefore, for safety, I suggest removing the final of `db` and  leaving reset `db` to null.

Reply:
If I want to keep the final, I think of a way：

Add a boolean type `closed` status  to `shuffledb.LevelDB` and `shuffledb.RocksDB` and set `closed` to true when `shuffledb.*DB`, then check `closed` is false before each api call and throw an `IllegalStateException` if check fail




1. The first way is add a `closed` status to

Reply:
If we want to keep the final, I think of a way：

Add a boolean type closed status to `shuffledb.LevelDB` and `shuffledb.RocksDB` and set closed to true when `shuffledb.*DB.close()`, then check closed is false before each api call and throw an `IllegalStateException` if check fail

Do you think this would be better? @mlaflamm

Reply:
> If we want to keep the final, I think of a way：
> 
> Add a boolean type closed status to `shuffledb.LevelDB` and `shuffledb.RocksDB` and set closed to true when `shuffledb.*DB.close()`, then check closed is false before each api call and throw an `IllegalStateException` if check fail
> 
> Do you think this would be better? @mridulm

@mridulm I test this way, it seems work, but since methods `removeAppAttemptPathInfoFromDB ` and `removeOldApplicationAttemptsFromDb ` catch all `Exception` now, so the above  test scenario will sill catch `RejectedExecutionException `, not `IllegalStateException`

Reply:
Or should I remove this test scenario in this pr.

```
The API correctness and availability of RemoteBlockPushResolver after calling the close method
```

which seems to be another issue

Reply:
@mridulm [0f2afe9](https://github.com/apache/spark/pull/37624/commits/0f2afe966336c63703cde1df6366db84d7a68ee9) keep `DB` as final and remove the following asserts:

```
 assert(ShuffleTestAccessor.mergeManagerLevelDB(resolver) == null)
 val message = intercept[RejectedExecutionException] {
      val appId = ApplicationId.newInstance(0, maxId)
      resolver.applicationRemoved(appId.toString, true)
    }.getMessage
assert(message.contains("Terminated, pool size = 0, " +
    s"active threads = 0, queued tasks = 0, completed tasks = $maxId"))
```

Already file a new jira [SPARK-40250](https://issues.apache.org/jira/browse/SPARK-40250) to further check the issue mentioned above.

Do you think this is OK?

Reply:
To clarify, as part of the tests we do check for the expected behavior (that the threadpool has been shutdown) ... What we removed is to check for a specific runtime exception being raised through a codepath which should not get invoked by NM (as the service is being shutdown).

That sounds fine to me.

Reply:
Yes, you understand right